src/Dependencies/Collections/SegmentedList`1.cs

```diff
@@ -135,12 +135,39 @@ public int Capacity
                 {
                     if (value > 0)
                     {
-                        var newItems = new SegmentedArray<T>(value);
-                        if (_size > 0)
+                        // Rather than creating a copy of _items, instead reuse as much of it's data as possible.
+                        // This saves as much as 50% of allocations and 70% of CPU cost of Add in large collections.
+                        // See SegmentedListBenchmarks_Add for repro details.
+                        var segments = SegmentedCollectionsMarshal.AsSegments(_items);
+
+                        var segmentSize = SegmentedArrayHelper.GetSegmentSize<T>();
+                        var segmentShift = SegmentedArrayHelper.GetSegmentShift<T>();
+                        var oldSegmentCount = segments.Length;
+                        var newSegmentCount = (value + segmentSize - 1) >> segmentShift;
+
+                        // Grow the array of segments, if necessary
+                        Array.Resize(ref segments, newSegmentCount);
+
+                        var lastPageSize = value - ((newSegmentCount - 1) << segmentShift);
+
+                        // If the previous last page is still the last page, resize it to lastPageSize.
+                        // Otherwise, resize it to SegmentSize.
+                        if (oldSegmentCount > 0)
                         {
-                            SegmentedArray.Copy(_items, newItems, _size);
+                            Array.Resize(
+                                ref segments[oldSegmentCount - 1],
+                                oldSegmentCount == newSegmentCount ? lastPageSize : segmentSize);
                         }
-                        _items = newItems;
+
+                        // Create all new pages (except the last one which is done separately)
+                        for (var i = oldSegmentCount; i < newSegmentCount - 1; i++)
+                            segments[i] = new T[segmentSize];
+
+                        // Create a new last page if necessary
+                        if (oldSegmentCount < newSegmentCount)
+                            segments[newSegmentCount - 1] = new T[lastPageSize];
```

Comment:
`Array.Resize` works even when the argument is null. This should handle all cases:

```
for (var i = oldSegmentCount - 1; i < newSegmentCount - 1; i++)
  Array.Resize(ref segments[i], segmentSize);

Array.Resize(ref segments[newSegmentCount - 1], lastPageSize);
```

Reply:
➡️ Now resolved